### ISSUE TITLE
Cannot use amplify-js in browser environment

### ISSUE THREAD

**Issue by sem4phor** [aws-amplify/amplify-js#7499]
**Describe the bug**

The package aws-amplify is meant to be used in the browser, isnt it? Therefore it should not rely on node builtins, which sadly is the case and prevents the usage of this package in many js build tools like rollup, snowpack or vitejs.

The @aws-cli package which is used by aws-amplify relies on various node build in modules like fs, process or http2, ....

I described the issue also here (https://github.com/vitejs/vite/issues/1374) and Evan You the creator of vuejs and vite mentioned that webpack will drop support for node build ins in version five. Because webpack is used by tools such as create-react-app  and vue-cli this will block many developers from using the amplify package soon I think!

**To Reproduce**
1. Download https://github.com/sem4phor/aws-cli-bug
2. Run npm run dev (node required)
3. See error in console

**Expected behavior**
The package works in the browser and does not import any node build ins which are not available in browser environment.

**Code Snippet**
https://github.com/sem4phor/aws-cli-bug

**Screenshots**
```
[plugin:vite:imports] Failed to resolve import "process". Does the file exist?
/Users/XXXXX/node_modules/@aws-sdk/util-user-agent-node/dist/es/index.js:1:22
1  |  import process from "process";
   |                       ^
2  |  export function defaultUserAgent(packageName, packageVersion) {
3  |      var engine = process.platform + "/" + process.version;
Error: Failed to resolve import "process". Does the file exist?
    at TransformContext.error (C:\Users\XXXXX\node_modules\vite\dist\node\chunks\dep-262f13eb.js:59552:50)
    at TransformContext.transform (C:\XXXXX\node_modules\vite\dist\node\chunks\dep-262f13eb.js:52957:30)
    at async Object.transform (C:\Users\XXXXX\node_modules\vite\dist\node\chunks\dep-262f13eb.js:59742:30)
    at async transformRequest (C:\Users\XXXXX\node_modules\vite\dist\node\chunks\dep-262f13eb.js:65977:29)
    at async C:\Users\XXXXX\node_modules\vite\dist\node\chunks\dep-262f13eb.js:66068:32
```

**What is Configured?**
Not relevant for this error.
 
- OS: Windows 10
- Browser: Chrome 87
- aws-amplify 3.3.13

**Comment by Jexordexan** [comment]
I'm experiencing the same issue, I have had no luck finding a workaround. I am also using `vite`. Also @sem4phor, in the issue body, you mention `@aws-cli` but I believe the problem is with `@aws-sdk`.

**Comment by soerenmartius** [comment]
I second that. The problem seems to be in `@aws-sdk`Seems like the library has dependencies on node built-in types which is really bad. Seems like the library was built for webpack and is not usable with vite or rollup at the moment? Also, the coupling increases the size of the package significantly.

Are there any plans to get rid of the node dependencies anytime soon?

**Comment by ebisbe** [comment]
I have the same issue

**Comment by Mootook** [comment]
I was also getting this issue when trying to migrate an existing project onto Vite.

Apparently work is being done on the aws-sdk to update this:
https://github.com/vitejs/vite/issues/1502#issuecomment-758822680 

But I had to follow this to get my app running as Vite is using rollup:
https://dev.to/swyx/making-aws-amplify-work-with-rollup-2d9m

**Comment by ebisbe** [comment]
> But I had to follow this to get my app running as Vite is using rollup:
> https://dev.to/swyx/making-aws-amplify-work-with-rollup-2d9m

@Mootook Did that link work for you? I tried and couldn't make it work. I'm new to Rollup / Vite so I wasn't even sure it would work. I'll try again if that worked then.

**Comment by mbp** [comment]
> > But I had to follow this to get my app running as Vite is using rollup:
> > https://dev.to/swyx/making-aws-amplify-work-with-rollup-2d9m
> 
> @Mootook Did that link work for you? I tried and couldn't make it work. I'm new to Rollup / Vite so I wasn't even sure it would work. I'll try again if that worked then.

I have the same issue (and also new to Rollup / Vite), so I am not sure how you would make the changes in e.g. `vite.config.ts` file where the syntax is different. Let me know if you make it work!

**Comment by sem4phor** [comment]
@Mootook It would be great if you could provide a repository with your suggested workaround :)

**Comment by ebisbe** [comment]
From what I have read the plugin interface of Vite is the same as Rollup so they should be compatible. What I don't understand is that in https://vite-rollup-plugins.patak.dev you can see that `Json` and `node-resolve` are already covered on Vite.

**Comment by Mootook** [comment]
@ebisbe, I didn't know that, good catch.

I have uploaded a small demo https://github.com/Mootook/aws-vite-demo/tree/main

But you're right, the plugin extension in `vite.config.js` is not necessary. Commenting it out, removing them as dependencies, and rebuilding seems to have no effect.

The only other issue I ran into was a global is not defined error, which has a solution in the index.html file.

**Comment by sammartinez** [comment]
@sem4phor Are you using Amplify JavaScript or using AWS SDK? I believe your issue is related to the AWS SDK instead of Amplify JS. Can you confirm?

**Comment by ebisbe** [comment]
@sammartinez Doesn't Aws-amplify use Aws-sdk? I use Aws-amplify but my errors come from node_modules/@aws-sdk and I have not installed in the package.json

**Comment by ebisbe** [comment]
> @ebisbe, I didn't know that, good catch.
> 
> I have uploaded a small demo https://github.com/Mootook/aws-vite-demo/tree/main
> 
> But you're right, the plugin extension in `vite.config.js` is not necessary. Commenting it out, removing them as dependencies, and rebuilding seems to have no effect.
> 
> The only other issue I ran into was a global is not defined error, which has a solution in the index.html file.

Your code is working indeed. But when I add your `vite.config.js` it still fails in mine... I've been trying to sample the minimum to make it work on my project but it doesn't. Besides what you have I'm using Tailwind

**Comment by sammartinez** [comment]
@ebisbe correct, we use AWS SDK but it is a dependency and separate team and [repo](https://github.com/aws/aws-sdk-js) that handles errors related to it.

**Comment by Mootook** [comment]
@ebisbe, I updated my repo to use tailwind and it's working.

What errors are you getting?
Is it still failing on the amplify dependency?

If not, I recommend moving this discussion because it could be something else.
there is also a vite specific discord for general troubleshooting
https://discord.gg/y44Ad4hF

**Comment by mbp** [comment]
@Mootook correct me if I am wrong, but I don't actually see you using aws-amplify in your repo (other than the package.json dependency)

If you insert the following code into `main.js`:

``` 
import Amplify, { Auth } from 'aws-amplify';

Amplify.configure({});
```  

Then `npm run build` will start failing with

```
'request' is not exported by __vite-browser-external, imported by node_modules\@aws-sdk\credential-provider-imds\dist\es\remoteProvider\httpRequest.js
``` 

This is the error message originally reported in the chain of issues: https://github.com/vitejs/vite/issues/1502 => https://github.com/vitejs/vite/issues/1374#issuecomment-754820938 => this

**Comment by Mootook** [comment]
@mbp 
Ah! Yes, you're correct. Apologies, I just went off the original steps to reproduce, which just included `npm run dev` (which does work, even with `Amplify.configure()`...odd), but the error is present in `npm run build`.

So credential provider is expecting the node http global which is not bundled with rollup. 
There is the 'rollup-plugin-node-builtins' package, but it seems outdated and sort of undercuts what Vite is striving for.

**Comment by ebisbe** [comment]
> @ebisbe, I updated my repo to use tailwind and it's working.
> 
> What errors are you getting?
> Is it still failing on the amplify dependency?
> 
> there is also a vite specific discord for general troubleshooting
> https://discord.gg/y44Ad4hF

It's funny because when I moved my code to your repo it's all working but in mine it still failing. I've forked your repo with my code added in https://github.com/ebisbe/aws-vite-demo and my failing code in here https://github.com/ebisbe/skedr-prototype/tree/live

The error I get is: 
```
⌊☻ yarn dev
yarn run v1.22.10
$ vite --open
Pre-bundling dependencies:
  vue
  vue-router
  aws-amplify
  luxon
  lodash/groupBy
  (...and 1 more)
(this will be run only when your dependencies or config have changed)
 > node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js: error: No matching export for import "request"
    4 │ import { request } from "http";
      ╵          ~~~~~~~

 > node_modules/@aws-sdk/shared-ini-file-loader/dist/es/index.js: error: No matching export for import "readFile"
    2 │ import { readFile } from "fs";
      ╵          ~~~~~~~~

 > node_modules/@aws-sdk/shared-ini-file-loader/dist/es/index.js: error: No matching export for import "homedir"
    3 │ import { homedir } from "os";
      ╵          ~~~~~~~

 > node_modules/@aws-sdk/shared-ini-file-loader/dist/es/index.js: error: No matching export for import "join"
    4 │ import { join, sep } from "path";
      ╵          ~~~~

 > node_modules/@aws-sdk/shared-ini-file-loader/dist/es/index.js: error: No matching export for import "sep"
    4 │ import { join, sep } from "path";
      ╵                ~~~

 > node_modules/@aws-sdk/credential-provider-process/dist/es/index.js: error: No matching export for import "exec"
    4 │ import { exec } from "child_process";
      ╵          ~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/write-request-body.js: error: No matching export for import "Readable"
    1 │ import { Readable } from "stream";
      ╵          ~~~~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http-handler.js: error: No matching export for import "Agent"
    3 │ import { Agent as hAgent, request as hRequest } from "http";
      ╵          ~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http-handler.js: error: No matching export for import "request"
    3 │ import { Agent as hAgent, request as hRequest } from "http";
      ╵                           ~~~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http-handler.js: error: No matching export for import "Agent"
    4 │ import { Agent as hsAgent, request as hsRequest } from "https";
      ╵          ~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http-handler.js: error: No matching export for import "request"
    4 │ import { Agent as hsAgent, request as hsRequest } from "https";
      ╵                            ~~~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http2-handler.js: error: No matching export for import "connect"
    4 │ import { connect, constants } from "http2";
      ╵          ~~~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/node-http2-handler.js: error: No matching export for import "constants"
    4 │ import { connect, constants } from "http2";
      ╵                   ~~~~~~~~~

 > node_modules/@aws-sdk/node-http-handler/dist/es/stream-collector/collector.js: error: No matching export for import "Writable"
    2 │ import { Writable } from "stream";
      ╵          ~~~~~~~~

 > node_modules/@aws-sdk/util-body-length-node/dist/es/index.js: error: No matching export for import "lstatSync"
    1 │ import { lstatSync } from "fs";
      ╵          ~~~~~~~~~

 > node_modules/@aws-sdk/eventstream-serde-node/dist/es/EventStreamMarshaller.js: error: No matching export for import "Readable"
    4 │ import { Readable } from "stream";
      ╵          ~~~~~~~~

 > node_modules/@aws-sdk/hash-stream-node/dist/es/hash-calculator.js: error: No matching export for import "Writable"
    2 │ import { Writable } from "stream";
      ╵          ~~~~~~~~

 > node_modules/@aws-sdk/hash-stream-node/dist/es/index.js: error: No matching export for import "createReadStream"
    1 │ import { createReadStream } from "fs";
      ╵          ~~~~~~~~~~~~~~~~

error when starting dev server:
Error: Build failed with 18 errors:
node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js:4:9: error: No matching export for import "request"
node_modules/@aws-sdk/credential-provider-process/dist/es/index.js:4:9: error: No matching export for import "exec"
node_modules/@aws-sdk/eventstream-serde-node/dist/es/EventStreamMarshaller.js:4:9: error: No matching export for import "Readable"
node_modules/@aws-sdk/hash-stream-node/dist/es/hash-calculator.js:2:9: error: No matching export for import "Writable"
node_modules/@aws-sdk/hash-stream-node/dist/es/index.js:1:9: error: No matching export for import "createReadStream"
...
    at failureErrorWithLog (/Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:1124:15)
    at buildResponseToResult (/Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:879:32)
    at /Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:974:20
    at /Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:541:9
    at handleIncomingPacket (/Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:624:9)
    at Socket.readFromStdout (/Volumes/Data/Development/skedr/next_prototype/node_modules/esbuild/lib/main.js:508:7)
    at Socket.emit (events.js:314:20)
    at addChunk (_stream_readable.js:298:12)
    at readableAddChunk (_stream_readable.js:273:9)
    at Socket.Readable.push (_stream_readable.js:214:10)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by sem4phor** [comment]
> Are you using Amplify JavaScript or using AWS SDK? I believe your issue is related to the AWS SDK instead of Amplify JS. Can you confirm?

@sammartinez I am using amplify which has aws-sdk as it's dependency

**Comment by sammartinez** [comment]
@sem4phor I guess I am a little confused and will need more information from you. What is your use case you are looking to do ? In looking at the code sample provided, thank you by the way it was helpful, I am not seeing a direct dependency of `aws-amplify` or an `aws-exports` file for testing the configuration, just an empty object passed. To some of the errors called out in the stack traces above, we have done some clean up on the library recently to remove unused dependencies. Can you look at testing the latest version of `aws-amplify` with your sample?

**Comment by madmod** [comment]
I too am having this issue. I have imported `@aws-sdk/client-cognito-identity` and `@aws-sdk/credential-provider-cognito-identity` directly. I am using Vite and not using Amplify fyi. Hopefully that helps you track down the root cause.

**Comment by madmod** [comment]
I have determined that the following modules cause build errors related to expecting various node js modules in the browser such as `http` and `fs`.
```
        '@aws-sdk/credential-provider-imds',
        '@aws-sdk/credential-provider-process',
        '@aws-sdk/shared-ini-file-loader',
        '@aws-sdk/eventstream-serde-node',
        '@aws-sdk/hash-node',
        '@aws-sdk/node-http-handler',
        '@aws-sdk/util-body-length-node',
        '@aws-sdk/util-user-agent-node',
```

**Comment by bhavinkamani-gt** [comment]
I am facing the same issue while using `vite`. 

I am getting the previously reported error `'request' is not exported by __vite-browser-external, imported by node_modules\@aws-sdk\credential-provider-imds\dist\es\remoteProvider\httpRequest.js`. Is this something that is fixed? I have tried this with the latest version of aws-amplify `3.3.22`.

I tried with rollup config as suggested earlier in this thread. I also experimented with `rollup-plugin-polyfill-node`. Both of them fails to fix the issue.

Here's my `vite.config.js`
```
import svelte from "rollup-plugin-svelte-hot";
import autoPreprocess from "svelte-preprocess";
import nodeResolve from "@rollup/plugin-node-resolve";
import json from "@rollup/plugin-json";

export default {
    alias: {
        "@": resolve(__dirname, "src"),
    },
    root: "src",
    plugins: [
        svelte({
            preprocess: autoPreprocess(),
            dev: true,
            hot: true,
        }),
    ],
    build: {
        rollupOptions: {
            plugins: [
                nodeResolve({
                    browser: true,
                    preferBuiltins: false,
                }),
                json(),
            ],
        },
    },
};
```

Error dump

```
'request' is not exported by __vite-browser-external, imported by node_modules/.pnpm/@aws-sdk/credential-provider-imds@3.6.1/node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
file: /temp/tst/node_modules/.pnpm/@aws-sdk/credential-provider-imds@3.6.1/node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js:4:9
2: import { ProviderError } from "@aws-sdk/property-provider";
3: import { Buffer } from "buffer";
4: import { request } from "http";
            ^
5: /**
6:  * @internal
error during build:
Error: 'request' is not exported by __vite-browser-external, imported by node_modules/.pnpm/@aws-sdk/credential-provider-imds@3.6.1/node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
    at error (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:5279:30)
    at Module.error (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:9996:16)
    at Module.traceVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:10389:29)
    at ModuleScope.findVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:8847:39)
    at FunctionScope.findVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:2641:38)
    at ChildScope.findVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:2641:38)
    at FunctionScope.findVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:2641:38)
    at ChildScope.findVariable (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:2641:38)
    at Identifier.bind (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:4017:40)
    at CallExpression.bind (/temp/tst/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:2728:23)
```

**Comment by jarrodu** [comment]
I am so glad I found this issue since I was about to write my own.

I am also having this issue.

**Comment by vovawed** [comment]
Hi @bhavinkamani-gt, have you manage to fix it?

**Comment by jarrodu** [comment]
I noticed that the "to-be-reproduced" tag was added to this issue. I am willing to spend a bit of time setting up a minimal example of the issue I am having. Is there a protocol for submitting an example or is a temporary public git repo with a README enough?

**Comment by bhavinkamani-gt** [comment]
@vovawed  No, I haven't yet found the solution.. As a workaround, I am using snowpack to do my production builds. Snowpack has a built-in support for node-polyfills with `snowpack build --polyfill-node`.

As commented [here](https://github.com/vitejs/vite/issues/1502#issuecomment-758822680). This should have been fixed in this library by isolating node related code. However, it doesn't seem to have happen. 

@sammartinez Would you have any update on this issue?

**Comment by sem4phor** [comment]
I think to be reprouced relates to a maintainer who needs to reproduce this. I have already submitted a reproduction repository...

Moreover the label Amplify UI Components is wrong, too IMO. The issue persists without importing ui components @sammartinez

**Comment by sammartinez** [comment]
Thanks for this feedback, I have removed the label. I will have my oncall look into this tomorrow.

**Comment by jarrodu** [comment]
When using Vite and the authentication documentation here: https://docs.amplify.aws/ui/auth/authenticator/q/framework/vue

I hit the same nodejs issues this issue describes.

So with this,
```
 "dependencies": {
    "@aws-amplify/ui-components": "^1.1.0",
    "aws-amplify": "^3.3.24",
    "vue": "^3.0.5"
  }
```
I get this,

`'request' is not exported by __vite-browser-external,`

**Comment by asmajlovicmars** [comment]
Having the same issue with SvelteKit and Amplify, and not including aws-sdk. Just Amplify js, but aws-sdk is called implicitly, as a dependency I guess... 

request' is not exported by __vite-browser-external, imported by node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
file: /Users/asmajlovic/code/sveltekit/p1/node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js:4:9
2: import { ProviderError } from "@aws-sdk/property-provider";
3: import { Buffer } from "buffer";
4: import { request } from "http";
            ^
5: /**
6:  * @internal
> 'request' is not exported by __vite-browser-external, imported by node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js

**Comment by ErikCH** [comment]
I found a workaround that can help for now. Eventually this will probably need to be fixed in the `@aws-sdk`

Add a vite.confg.js file to the root of the project. You'll need to add a new resolve.alias.

```
// vite.config.js
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';

// https://vitejs.dev/config/
export default defineConfig({
        plugins: [vue()],
        resolve: {
                alias: [
                  {
                        find: "./runtimeConfig",
                        replacement: ("./runtimeConfig.browser"),
                  },
                ],
          },

});
```

Then inside the index.html file add a script tag below the `<div id="app></div>` but above the module script import

```
<!-- index.html -->
<script>
      window.global = window;
      var exports = {}
</script>
```

You need to set `window.global` and `exports`. If you don't set `window.global` it won't work in development. If you don't add in `var exports = {}` then it will build, but will throw an error when you open it up on your web browser. 

I made a branch here you can take a look at too. https://github.com/ErikCH/amplify-js/tree/hotfix/error-with-vite

**Comment by bhvngt** [comment]
Thanks a ton @ErikCH. It works good for now. My builds are working without a hitch. 

As mentioned, will look forward to the long term solution.

**Comment by rafrasenberg** [comment]
Awesome!! Many thanks @ErikCH , finally a solution that works. Legend!

**Comment by josephfh** [comment]
Thanks a lot for the solution @ErikCH I can't believe my luck that you posted this today!

**Comment by asmajlovicmars** [comment]
Hi @ErikCH, I tried to do a similar thing with SvelteKit, but no luck yet. Vite.js config is stored in svelte.config.cjs:

`
// svelte.config.cjs
const st = require('@sveltejs/adapter-static');

module.exports = {
  kit: {
    adapter: st(),
    target: '#svelte',
    vite: {
      build: {
        minify: false,
      },
      resolve: {
        alias: [
          {
            find: './runtimeConfig',
            replacement: './runtimeConfig.browser',
          },
        ],
      },
    },
  },
};
`

Error: 

`
vite v2.1.2 building for production...
✓ 741 modules transformed.
'request' is not exported by __vite-browser-external, imported by node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
file: /Users/asmajlovic/code/sveltekit/p1/node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js:4:9
2: import { ProviderError } from "@aws-sdk/property-provider";
3: import { Buffer } from "buffer";
4: import { request } from "http";
            ^
5: /**
6:  * @internal
> 'request' is not exported by __vite-browser-external, imported by node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
Error: 'request' is not exported by __vite-browser-external, imported by node_modules/@aws-sdk/credential-provider-imds/dist/es/remoteProvider/httpRequest.js
    at error (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:5307:30)
    at Module.error (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:9716:16)
    at Module.traceVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:10104:29)
    at ModuleScope.findVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:8874:39)
    at FunctionScope.findVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:2657:38)
    at ChildScope.findVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:2657:38)
    at FunctionScope.findVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:2657:38)
    at ChildScope.findVariable (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:2657:38)
    at Identifier.bind (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:4042:40)
    at CallExpression.bind (/Users/asmajlovic/code/sveltekit/p1/node_modules/rollup/dist/shared/rollup.js:2744:23)
`

**Comment by ErikCH** [comment]
@asmajlovicmars In your `svelte.config.cjs` import your `resolve.alias` like this

```
resolve: {
        alias: {
          "./runtimeConfig": "./runtimeConfig.browser"
        }
      }
```

**Comment by asmajlovicmars** [comment]
Thanks a lot @ErikCH,
Your solution unblocked the SvelteKit + Amplify build!

However, when I tried to run the built code, which works perfectly with **npm run dev**, it gets stuck when it gets to Amplify. If it helps, I built a small repo, which only needs a valid aws-exports.js to test this. 

Repo: https://github.com/adnansmajlovic/svk
Error: ReferenceError: AmplifyClass is not defined

Of course, I'm still looking into it, but if anything comes to mind, please share. 

Super thankful for unblocking the build 🎉

**Comment by ErikCH** [comment]
@asmajlovicmars Looking at the repository, I saw some old `@aws-amplify` imports. You may want to double check those and switch them to named imports from version 3 `import { Analytics } from 'aws-amplify';`.

I did some tests and I might have found another work around for SveltKit. To run in dev mode, take out ssr inside the config. 
```
// remove this
ssr: {
	noExternal: Object.keys(pkg.dependencies || {})
	},
``` 

When I tested this allowed me to work in dev mode. 

However, to build, I had to add it back in SSR. I also added in a few extra configurations into the config file, since it didn't seem to pick them up from the HTML file.
Here is the vite portion of the config. 

```
...
vite: {
      ssr: {
        noExternal: Object.keys(pkg.dependencies || {})
      },
      resolve: {
        alias: {
          "./runtimeConfig": "./runtimeConfig.browser"
        }
      },
      build: {
        rollupOptions: {
          output: {
            intro: "if(exports === undefined){var exports ={}; var self = {}}"
          }
        }
      }
    }
...
```

With this configuration I was able to use Auth, after building and running it.

**Comment by asmajlovicmars** [comment]
Thank you @ErikCH! 

It finally builds with Amplify, and as you confirmed Auth works too! I used old imports only because I had to import them dynamically in onMount, since the regular import wouldn't work, but now everything is as it should be, thanks to you! 😁 🎉 

Will keep looking if it's possible to have dev and build mode working without changing ssr in svelte.config.cjs, but that's perfectly fine! 

Once again, thanks a lot for your help!!!  🙌

**Comment by TheDutchCoder** [comment]
You can also shim this by using this pattern: https://github.com/Azure/azure-sdk-for-js/issues/14496#issuecomment-808442337

**Comment by radiosilence** [comment]
Am using pure Vite+React and whilst the above solved the build, I now get this in browser:
![Screenshot 2021-04-30 at 10 12 19](https://user-images.githubusercontent.com/65021/116677927-ad29ee80-a9a0-11eb-8307-f95a31f73fa5.png)

:(

**Comment by ErikCH** [comment]
I was able to reproduce that as well @radiosilence .

A quick work around is don't use the `@aws-amplify/ui-react` library. Use the `aws-amplify` library instead. You'll have to create your own UI, but it should work. Here is the guide to the API https://docs.amplify.aws/lib/q/platform/js

**Comment by signal-intrusion** [comment]
My silly workaround for a static site with some TypeScript, but without Webpack was to import the entire library into my "main" file: 
```
// app.ts
import * as amplify from 'aws-amplify/dist/aws-amplify.min.js'

amplify.Auth.configure() // you get the idea
```

**Comment by simlevesque** [comment]
> Am using pure Vite+React and whilst the above solved the build, I now get this in browser:

I get the same error, this seems to me like it's a StencilJS error.

**Comment by tranthaison1231** [comment]
@simlevesque I have the same error live you, do you have any idea to fix it??

**Comment by simlevesque** [comment]
@tranthaison1231

did you try the fix I mentionned in my previous comment ?

**Comment by tranthaison1231** [comment]
![image](https://user-images.githubusercontent.com/43759215/120258277-eadda800-c2bb-11eb-939b-316255e373f1.png)
I got an error like this.

**Comment by half2me** [comment]
> @asmajlovicmars In your `svelte.config.cjs` import your `resolve.alias` like this
> 
> ```
> resolve: {
>         alias: {
>           "./runtimeConfig": "./runtimeConfig.browser"
>         }
>       }
> ```

thanks this worked for me

**Comment by radiosilence** [comment]
@ErikCH Honeslty I eventually solved this by getting rid of Amplify and just writing a light wrapper around `amazon-cognito-identity-js` and my own login form.


Here it is if anyone wants to use it:

```ts
import * as AmazonCognitoIdentity from "amazon-cognito-identity-js";

let userPool: AmazonCognitoIdentity.CognitoUserPool;

export function configure(
  poolConfig: AmazonCognitoIdentity.ICognitoUserPoolData
) {
  userPool = new AmazonCognitoIdentity.CognitoUserPool(poolConfig);
}

function checkPool() {
  if (!userPool) throw new Error("Cognito User Pool has not been configured");
}

export async function getSession(): Promise<AmazonCognitoIdentity.CognitoUserSession> {
  checkPool();
  return new Promise((resolve, reject) => {
    const currentUser = userPool.getCurrentUser();
    if (!currentUser) return reject("not signed in");
    currentUser.getSession(
      (err: unknown, session: AmazonCognitoIdentity.CognitoUserSession) => {
        if (err) return reject(err);
        if (!session) return reject("no user data");
        resolve(session);
      }
    );
  });
}

export async function getValidSessionUser(): Promise<AmazonCognitoIdentity.CognitoUser> {
  checkPool();
  return new Promise((resolve, reject) => {
    const currentUser = userPool.getCurrentUser();
    if (!currentUser) return reject("not signed in");
    currentUser.getSession(
      (err: unknown, session: AmazonCognitoIdentity.CognitoUserSession) => {
        if (err) return reject(err);
        if (!session) return reject("no session");
        if (!session.isValid) return reject("session invalid");
        resolve(currentUser);
      }
    );
  });
}

export async function getUserData(): Promise<AmazonCognitoIdentity.UserData> {
  checkPool();
  const currentUser = await getValidSessionUser();
  return new Promise((resolve, reject) => {
    currentUser.getUserData((err, userData) => {
      if (err) return reject(err);
      if (!userData) return reject("no user data");
      resolve(userData);
    });
  });
}

export async function getUserAttributes(): Promise<Record<string, string>> {
  checkPool();
  const currentUser = await getValidSessionUser();
  return new Promise((resolve, reject) => {
    currentUser.getUserAttributes((err, userAttributes) => {
      if (err) return reject(err);
      if (!userAttributes) return reject("no user data");
      resolve(
        userAttributes.reduce(
          (acc: Record<string, string>, attribute) => ({
            ...acc,
            [attribute.Name]: attribute.Value,
          }),
          {}
        )
      );
    });
  });
}

export async function getIdToken() {
  checkPool();
  const session = await getSession();
  if (!session.isValid) throw new Error("invalid session");
  return session.getIdToken().getJwtToken();
}

export function signIn(username: string, password: string) {
  checkPool();
  return new Promise((resolve, reject) => {
    const authenticationDetails =
      new AmazonCognitoIdentity.AuthenticationDetails({
        Username: username,
        Password: password,
      });
    const cognitoUser = new AmazonCognitoIdentity.CognitoUser({
      Username: username,
      Pool: userPool,
    });
    cognitoUser.authenticateUser(authenticationDetails, {
      onSuccess: (result) => resolve(result),
      onFailure: (error) => reject(error),
      // newPasswordRequired: () => reject(new Error('not implemented: new password required')),
      newPasswordRequired: () => {
        cognitoUser.completeNewPasswordChallenge(
          password,
          {},
          { onSuccess: resolve, onFailure: reject }
        );
      },
      mfaRequired: () => reject(new Error("not implemented: mfa required")),
      totpRequired: () => reject(new Error("not implemented: totp required")),
      customChallenge: () =>
        reject(new Error("not implemented: custom challenge")),
      mfaSetup: () => reject(new Error("not implemented: mfa setup")),
      selectMFAType: () =>
        reject(new Error("not implemented: select mfa type")),
    });
  });
}

export function forgotPassword(username: string) {
  checkPool();
  return new Promise((resolve, reject) => {
    const cognitoUser = new AmazonCognitoIdentity.CognitoUser({
      Username: username,
      Pool: userPool,
    });
    cognitoUser.forgotPassword({
      onSuccess: (result) => resolve(result),
      onFailure: (error) => reject(error),
    });
  });
}

export function confirmPassword(
  username: string,
  verificationCode: string,
  newPassword: string
) {
  checkPool();
  return new Promise((resolve, reject) => {
    const cognitoUser = new AmazonCognitoIdentity.CognitoUser({
      Username: username,
      Pool: userPool,
    });
    cognitoUser.confirmPassword(verificationCode, newPassword, {
      onSuccess: () => resolve(null),
      onFailure: (error) => reject(error),
    });
  });
}

export function signOut() {
  checkPool();
  const currentUser = userPool.getCurrentUser();
  if (currentUser) {
    currentUser.signOut();
  }
}

```

**Comment by siegerts** [comment]
ping

**Comment by jarrodu** [comment]
Please don't close this ticket without fixing the underlining issue.

**Comment by mgerring** [comment]
I have the same issue as @radiosilence -- looks like there are workarounds for everything except for `@aws-amplify/ui-react`. Replacing the authentication UI components isn't in scope for us for now, so we're stuck for the time being.

**Comment by radiosilence** [comment]
@mgerring We ended up replacing it (and eventually replacing Amplify), it isn't a hugely horrible thing to do, took us about a day though we don't cover all states. There are a many other reasons other than this ticket to not use `@aws-amplify/ui-react` and also Amplify entirely.

The sooner you migrate to a lightweight promise based wrapper around the parts of `amazon-cognito-identity-js` you actually need, the sooner your headaches will disappear :)

**Comment by charles-stuart-appdetex** [comment]
@radiosilence is `amazon-cognito-identity-js` deprecated though, or did it just get moved? I find varying information on it.

**Comment by armenr** [comment]
# Temporary workaround that WORKS!

Howdy folks! I've got a working setup for aynone that's trying to get `@aws-amplify/ui-components` working with VITE + Vue3 - **_but it's just slightly hacky_**, and I stumbled upon it TOTALLY by chance.

Wanna give a shout out to @ErikCH for:

```
        alias: {
          "./runtimeConfig": "./runtimeConfig.browser"
        }
```

Wanna give another shoutout to @simlevesque for: 

```
import { defineCustomElements } from '@aws-amplify/ui-components/dist/components/index.js';
defineCustomElements(window);
```

Are you ready kids?! (SpongeBob reference - feel free to disregard).

## Notes before you dive in

I am working off of an increasingly popular (and honestly AMAZING) starter template called [`Vitesse`](https://github.com/antfu/vitesse)

Why do I love vitesse? (recommend you go to the repo and see the readme for more extensive details)
- really solid plugins included (and they are well supported/active)
- extremely flexible
- extremely fast
- extremely *productive*
- Supports SSR, SPA, and SSG...all with zero-config PWA (out of the box, no configs required)
- +1 --> I got it working with Amplify

### Amplify-related caveats of Vitesse that I've discovered

- Use YARN in your build pipeline instead of pnpm (easy to do by including your own `amplify.yml` file in your repo)

# Solution

## Step 1: `vite.config.ts`

You need to add the the `./runtimeConfig` alias

```
  resolve: {
    alias: {
      '~/': `${path.resolve(__dirname, 'src')}/`,
      './runtimeConfig': './runtimeConfig.browser',
    },
  },
```

You need to add vue --> template --> compiler options

```
      template: {
        compilerOptions: {
          isCustomElement: tag => tag.startsWith('amplify-'),
          sourceMap: true,
        },
      },
```

Added this, but not sure if it's needed (please help me test if you'd like!):

```
// In your imports section

import resolve from '@rollup/plugin-node-resolve'

// in your vite defineConfig --> plugins section
    resolve({
      browser: true,
      preferBuiltins: false, // new!
    }),
```

My entire config --> 

```
import path from 'path'
import { defineConfig } from 'vite'
import Vue from '@vitejs/plugin-vue'
import Markdown from 'vite-plugin-md'
import Pages from 'vite-plugin-pages'
import Prism from 'markdown-it-prism'
import { VitePWA } from 'vite-plugin-pwa'
import WindiCSS from 'vite-plugin-windicss'
import Layouts from 'vite-plugin-vue-layouts'
import resolve from '@rollup/plugin-node-resolve'
import ViteComponents, { HeadlessUiResolver /* PrimeVueResolver */ } from 'vite-plugin-components'
import VueI18n from '@intlify/vite-plugin-vue-i18n'
import ViteIcons, { ViteIconsResolver } from 'vite-plugin-icons'

// @ts-expect-error missing types
import LinkAttributes from 'markdown-it-link-attributes'

const markdownWrapperClasses = 'prose prose-sm m-auto text-left'

// ViteConfig
export default defineConfig({
  resolve: {
    alias: {
      '~/': `${path.resolve(__dirname, 'src')}/`,
      './runtimeConfig': './runtimeConfig.browser',
    },
  },

  // Load plugins
  plugins: [
    resolve({
      browser: true,
      preferBuiltins: false, // new!
    }),

    Vue({
      include: [/\.vue$/, /\.md$/],
      template: {
        compilerOptions: {
          isCustomElement: tag => tag.startsWith('amplify-'),
          sourceMap: true,
        },
      },
    }),

    // https://github.com/hannoeru/vite-plugin-pages
    Pages({
      extensions: ['vue', 'md'],
    }),

    // https://github.com/JohnCampionJr/vite-plugin-vue-layouts
    Layouts(),

    // https://github.com/antfu/vite-plugin-md
    Markdown({
      wrapperClasses: markdownWrapperClasses,
      headEnabled: true,
      markdownItSetup(md) {
        // https://prismjs.com/
        md.use(Prism)
        md.use(LinkAttributes, {
          pattern: /^https?:\/\//,
          attrs: {
            target: '_blank',
            rel: 'noopener',
          },
        })
      },
    }),

    // https://github.com/antfu/vite-plugin-components
    ViteComponents({
      // allow auto load markdown components under `./src/components/`
      extensions: ['vue', 'md'],

      // allow auto import and register components used in markdown
      customLoaderMatcher: id => id.endsWith('.md'),

      // generate `components.d.ts` for ts support with Volar
      globalComponentsDeclaration: true,

      // auto import icons
      customComponentResolvers: [
        // PrimeVueResolver(),
        HeadlessUiResolver(),
        // https://github.com/antfu/vite-plugin-icons
        ViteIconsResolver({
          componentPrefix: '',
          // enabledCollections: ['hero'],
        }),
      ],
    }),

    // https://github.com/antfu/vite-plugin-icons
    ViteIcons(),

    // https://github.com/antfu/vite-plugin-windicss
    WindiCSS({
      safelist: markdownWrapperClasses,
    }),

    // https://github.com/antfu/vite-plugin-pwa
    VitePWA({
      registerType: 'autoUpdate',
      includeAssets: ['favicon.svg', 'robots.txt', 'safari-pinned-tab.svg'],
      manifest: {
        name: 'Beep',
        short_name: 'BeepBeep',
        theme_color: '#ffffff',
        icons: [
          {
            src: '/pwa-192x192.png',
            sizes: '192x192',
            type: 'image/png',
          },
          {
            src: '/pwa-512x512.png',
            sizes: '512x512',
            type: 'image/png',
          },
          {
            src: '/pwa-512x512.png',
            sizes: '512x512',
            type: 'image/png',
            purpose: 'any maskable',
          },
        ],
      },
    }),

    // https://github.com/intlify/vite-plugin-vue-i18n
    VueI18n({
      runtimeOnly: true,
      compositionOnly: true,
      include: [path.resolve(__dirname, 'locales/**')],
    }),
  ],

  // Server conifg
  server: {
    fs: {
      strict: true,
    },
  },

  // Build settings
  // build: {
  //   polyfillDynamicImport: true,
  //   sourcemap: true,
  //   rollupOptions: {
  //     shimMissingExports: true,
  //     output: {
  //       manualChunks: undefined,
  //       inlineDynamicImports: true,
  //     },
  //   },
  // },

  // https://github.com/antfu/vite-ssg
  ssgOptions: {
    script: 'async',
    formatting: 'minify',
  },

  // OptimizeDeps build settings
  optimizeDeps: {
    include: ['vue', 'vue-router', '@vueuse/core'],
    exclude: ['vue-demi'],
  },
})
```

## `./modules/Amplify.ts`

So, in Vitesse, there's a concept of "modules" that is similar to "boot" files in the Quasar framework.

I created a "module" to load up Amplify on App boot, and this is where I use @simlevesque's import workaround

```
import Amplify from 'aws-amplify'

import {
  applyPolyfills,
  // defineCustomElements,
} from '@aws-amplify/ui-components/loader'

import { defineCustomElements } from '@aws-amplify/ui-components/dist/components/index'

// @ts-expect-error missing types
import awsmobile from './../aws-exports'
import { UserModule } from '~/types'

export const install: UserModule = async({ app }) => {
  // NOTE: "await" might not be doing anything useful here, at all
  // TODO: Revisit/refactor this
  await Amplify.configure(awsmobile)
  await applyPolyfills()
  await defineCustomElements(window)

  // NOTE: Uncertain if this is necessary
  // TODO: Revisit, test, and remove if not necessary
  app.config.compilerOptions.isCustomElement = tag => tag.startsWith('amplify-')
}
```
## Still not working! What happens next?

Once you get THIS far, you'll see two IMPORTANT errors in your console/terminal

### Error in console/terminal refers to `@stenciljs/core` missing a require/export

We install `@stencil/core` directly into our project -->  `pnpm install @stencil/core`

### Error in console/terminal refers to `zen-observable` missing a `default export`
`pnpm install https://github.com/abdonrd/zen-observable`

Here, we're installing a fork for zen-observable directly from a fork on GitHub that has a pending PR that the zen-observable maintainers are ignoring (the PR for this branch is neglected, but it works!)

### Final steps --> clear out your node_modules and re-install everything

3. `rm -rf node_modules` in your project or your monorepo package folder
4. `pnpm install` or `yarn` or whatever package manager you're using

## Results:

![Screen Shot 2021-08-01 at 3 08 35 PM](https://user-images.githubusercontent.com/7662191/127786704-b7326f4c-1624-4dac-a320-2afc9a372232.png)

**Comment by ErikCH** [comment]
@armenr Awesome! Thanks for the great work around!

**Comment by armenr** [comment]
@ErikCH and for anyone else who might find it useful --> here's a reproducible sample repo

https://github.com/armenr/amplify-vitesse-repro
